= Test inside the C / C++ sources
:toc:
:sectnums:
:sectlinks:
:cpp: C++
:cp: C/++


The test has gotten some ideas from the Google test framework (see link:https://chromium.googlesource.com/external/github.com/pwnall/googletest/+/refs/tags/release-1.8.0/googletest/docs/Primer.md[]) but is more easier.

The primary goal is, control a proper test output to a file while the program is running. This needs organization.

All (very less) sources are contained in `emC/Test/*`, especially `emC/Test/testAssert.h`. 

== Usage


For test any operation (C-function) can be used. It is not necessary to mark it as "TEST" operation. Usual you have a `main()` and in the `main()` you invoke some operations one after another, which executes some tests. 


=== Frame for test

Inside an operation the begin and the end of the test is marked with:

----
int anyOperation(int mayhaveArgs) {
  STACKTRC_ENTRY("anyOperation");
  TEST_TRY("Description of the test case") {
    //
    //... any stuff to test
    //
  }_TEST_TRY_END
  STACKTRC_LEAVE;
}
----
  
The macros `STACKTRC_...` are necessary for exception messaging with stacktrace-report, see link:../Base/ThCxtExc_emC.html[]. 

The `TEST_TRY(...)` macro declares a `bTESTok` variable and notice the start. For PC usage this macro is defined as 

 #define TEST_TRY(MSG) TRY { \
   bool bTESTok = true; \
   msgStartFileLine_testAssert_emC(MSG, __FILE__, __LINE__);

The compiler intrinsic `FILE` and `LINE` writes the file and line information in the test output, important for tracking of test results in the sources. 

The `_TEST_TRY_END` checks the `bTESTok` variable and writes an proper message or stores any data information for the test result. It contains also the exception handling for unexpected exceptions while test. For PC usage this macro is defined as 

  #define _TEST_TRY_END  \
    msgEndFileLine_testAssert_emC(bTESTok); } _TRY  \
  CATCH(Exception, exc) { \
    exceptionFileLine_testAssert_emC(exc, __FILE__, __LINE__); \
  } END_TRY

For exceptions see link:../Base/ThCxtExc_emC.html[]. It an unexpected exception occurs during the test, then the `msgEndFileLine_testAssert_emC(...)` routine is not reached, instead the `exceptionFileLine_testAssert_emC(...)` routine is invoked. It writes the exception with file and line and if possible (depending on setting of `DEF_ThreadContext_STACKTRC`) it writes the stack trace to evaluate where and why the exception was occurring.  

That is all for the test frame.

Of course selecting some test cases with a specific tool (gradle tests or such) is not possible. But there is a normal programming for test cases. Selection of specific test cases can be done with ordinary conditional compiling. It doesn't need special knowledge. 

=== Execute tests

The tests itself checks and logs results. For example the test in `emc_Test_ObjectJc/test_ObjectJc.c`:

  bool bOk = CHECKstrict_ObjectJc(&data->base.obj
             , sizeof(myData1const), refl_MyType_Test_ObjectJc, 0);
  CHECK_TRUE(bOk, "checkStrict_ObjectJc with refl and without instance id.")

Any boolean test result condition is built. Then the `CHECK_TRUE` with the test result is invoked, with a text which describes the *positive* expected result. The `CHECK_TRUE` produces only an output if the check is false: 

 ERROR: checkStrict_ObjectJc with refl and without instance id. 
                           (@38: emC_Test_ObjectJc\test_ObjectJc.c)

in one line. Instead using `TEST_TRUE` outputs also the positive test:

 ok: checkStrict_ObjectJc with refl and without instance id.

This documents which tests are done. The output is valid for PC test or for test on an embedded platform with text output capability. 

The test can also output values, in a `printf` manner. For example in `emC_Test_Math/Test_Math_emC.c` the results of mathematics are checked:

 TEST_TRUE(rs == (int32)values[ix].rs
          , "muls16_emC %4.4X * %4.4X => %8.8X"
          , (int)(as) & 0xffff, (int)bs & 0xffff, rs);

It outputs:

  ok: muls16_emC 4000 * FFFF => FFFFC000
  ok: mulu16_emC 4000 * FFFF => 3FFFC000
  ok: muls16_emC 7FFF * FFFF => FFFF8001
  ....
  
and informs about numeric test result, here for a fix point multiplication. The test is important because the fix point multiplication can fail on an embedded platform if register widths etc. are other as expected. Maybe using `CHECK_TRUE` may be more proper, information only on faulty results. The calculated faulty value can be used to detect the reason of the fault.  

Furthermore the text can be prepared and reused for more test outputs. See the example testing a smoothing control block (T1) in `test1_T1_Ctrl_emC()` (`src/test/cpp/emC_Test_Ctrl/test_T1i_Ctrl.c`) the output of the control block for a step response is compared with the known mathematic result. The maximum of the abbreviation (error) is built in a variable `dysh1_eMax`. A small abbreviation is expectable because the T1 FBlock has only integer accuracy with a defined step width. It means in comparison to the exact mathematic behaviour not in the step time a small difference is expected and admissible. Hence the test is written as 

     CHECK_TRUE(dysh1_eMax < thiz->abbrTsh1, checktext[0]); 

The abbreviation is compared with a given value. Because some other tests are done too, the text is stored in an array which is used for more tests. Example: 

    char checktext[8][100];
    snprintf(checktext[0], 100, "%1.3f < abbreviation of Tsh1 4 Bit fTs", thiz->abbrTsh1);
    snprintf(checktext[1], 100, "%1.3f < abbreviation of Tsh2 8 Bit fTs", thiz->abbrTsh2);
    ....
    
You see, the functionality what and which to test is complex due to the technical approach. But the building of the test result is very simple. The `CHECK_TRUE(COND, TEXT,...)` macro expects only a `false` or `true` as first argument, `TEXT` as second arguemnt which describes the expected behavior, and maybe some more variable arguments. 

=== The test output

For test on PC or test on a text supporting embedded platform a text output is created.

With distinguish `TEST_TRUE` and `CHECK_TRUE` one can add more information about executed tests. A test output with some executed tests looks like:

----
Test: test_ctor_ObjectJc: (emC_Test_ObjectJc/test_ObjectJc.c @ 89) ...
  ok: refl type is ok
  ok: INIZ_VAL_MyType_Test_ObjectJc
  ok: checkStrict_ObjectJc
ok
----

You see the test case starts with `Test: ...` left aligned, and `ok` after the test is also written left aligned as the finish line. Between them some messages `  ok: ...` which documents which tests are executed, or `  ERROR: ...` if a test fails:

----
Test: test_ctor_ObjectJc: (emC_Test_ObjectJc\test_ObjectJc.c @ 89) ...
  ok: refl type is ok
  ERROR: INIZ_VAL_MyType_Test_ObjectJc (@99: emC_Test_ObjectJc\test_ObjectJc.c)
  ok: checkStrict_ObjectJc
ERROR
----

The simple form looks like:

----
Test: test_cos16: (emC_Test_Math\Test_Math_emC.c @ 228) ...
ok
----

If during processing the test algorithm an exception is thrown, then (using the `TEST_TRY(...)` macros) this test is aborted with an error message:

----
Test: test_ObjectJcpp_Base: (emC_Test_ObjectJc\test_ObjectJcpp.cpp @ 109) ...
EXCEPTION 1 (24, 0) @21: ..\..\src\test\cpp\emC_Test_ObjectJc\test_ObjectJcpp.cpp

RuntimeException: : 24=0x00000018
  at THROW (..\..\src\test\cpp\emC_Test_ObjectJc\test_ObjectJcpp.cpp:21)
  at test_ObjectJcpp_Base (..\..\src\test\cpp\emC_Test_ObjectJc\test_ObjectJcpp.cpp:109)
  at test_ObjectJcpp (..\..\src\test\cpp\emC_Test_ObjectJc\test_ObjectJcpp.cpp:214)
  at main (..\..\src\test\cpp\emC_TestAll\testBasics.cpp:21)
----

=== Exceptions while test

This exception in the chapter above had occured because the macro for `INIZsuper_ClassJc(...)` was faulty, not all elements are initialized. The type is tested inside a {Cpp} constructor outside of the test itself, and that causes the exception. For that case the test is finished with the `EXCEPTION ...` line. Additinal the stack trace is outputed. With that the source of the exception was able to found without elaborately debug tests: In line 21 an `ASSERT_emC(...)` was found, which checks the base type. Setting a breakpoint there (Visual Studio) shows, the information about the `ClassJc...superClass` was missing, which was caused by the faulty macro for the initialization. 

It is also possible to write `TRY { .... }_TRY CATCH { .... } ` statements inside the test, to catch an exception in the test algorithm. Then the CATCH block should contain: 

----
  CATCH(Exception, exc) {
    TEST_EXC(exc);
  }
----

This logs the not expected exception for the test output. But if the exception behaviour is tested itself as test case, it can be written (see `emC_Test_Stacktrc_Exc/TestException.cpp`):

----
  TRY{
    //raise(SIGSEGV);
    CALLINE; float val = testThrow(thiz, 10, 2.0f, _thCxt);
    printf("val=%f\n", val);
  }_TRY
  CATCH(Exception, exc) {
    CHECK_TRUE(exc->line == 46, "faulty line for THROW");
    bHasCatched = true;
  }  
  FINALLY {
    bHasFinally = true;
  } END_TRY;
  TEST_TRUE(bHasCatched && bHasFinally, "simple THROW is catched. ");
----

In the `CATCH` clause a test can also confirm, but to check whether the `CATCH` and also here `FINALLY` has entered, a boolean variable is set and test after the `TRY` block. 



The evaluation can be done with a not sophisticated text analyse, see link:../TestOrg/testStrategie_en.html#trueview-of-test-results[]



[#TestEmbdNum]
== Test environment in embedded control without text output

The macros can be written to store only values instead output the texts via printf. 

Either the text result are visit via debugger access after all tests, or via any data transfer.

TODO describe an example. 

The complete test for embedded control compiles for the target, load the target, start, and reads out the result via the communication line which may be JTAG or usual a serial connection. 


== How does it work


The macro is defined for PC usage as:

 /**Checks only, output only if error*/
 #define CHECK_TRUE(COND, MSG, ...) \
 if(!(COND)) { \
   expectMsgFileLine_testAssert_emC(false, MSG, __FILE__, __LINE__, ##__VA_ARGS__); \
   bTESTok = false; }

The routine `expectMsgFileLine_testAssert_emC(...)` writes the information about file and line in the error case, important for tracking of test results in the sources.

The `CHECK_TRUE(...)` macro can have additional values, the MSG should have placeholder like `%d` (as in printf) for that. Then values from the test can be outputted too. It may be helpfull to detect the error from outside doing a single step.  

If seen in the macro, it `COND` is true, the macro does nothing. If `CONF==false` the `bTESTok` variable is set to `false`. Hence the test is not ok, written on `TEST_END` 
 
There is a second macro:

 /**Test, output ok MSG if ok, only on error with file and line. */
 #define TEST_TRUE(COND, MSG, ...) \
 if(!expectMsgFileLine_testAssert_emC(COND, MSG, __FILE__, __LINE__, ##__VA_ARGS__)) \
   bTESTok = false;

This macro calls anyway the `expectMsg...(...)` routine. This routine notice the test and writes the `MSG` without file and line if the test is proper if `COND==true`. On `CONF==false` it writes the same as `CHECK_TRUE(...)`



== Test environment, mock, dependency injection

(additonal content with common meaning, TODO)

The test routines itself calls one or some routines from the module sources 
in an environment arranged in the respective test routine. If instances are necessary,
they are created and removed after test in the test routine. If additional depending
complex modules are necessary, they should be replaces by mock objects because elsewhere
the other module is tested too in a complex non-independent kind. The mock object
should be simple and can contain some helper for checking the test behavior. 
The possible usage of dependency injection instead instantiating of composite objects
inside the test object is a problem of the module source, not a problem of the test itself.

 
